vm/state_migration/src/nv12/mod.rs

```diff
@@ -0,0 +1,130 @@
+    actors_root_in: Cid,
+    prior_epoch: ChainEpoch,
+) -> MigrationResult<Cid> {
+    let mut jobs = FuturesOrdered::new();
+
+    // Maps prior version code CIDs to migration functions.
+    let mut migrations: HashMap<Cid, Rc<dyn ActorMigration<BS>>> =
+        HashMap::with_capacity(ACTORS_COUNT);
+    migrations.insert(
+        *actorv3::ACCOUNT_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::ACCOUNT_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::CRON_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::CRON_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::INIT_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::INIT_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::MULTISIG_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::MULTISIG_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::PAYCH_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::PAYCH_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::REWARD_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::REWARD_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::MARKET_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::MARKET_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::MINER_ACTOR_CODE_ID,
+        miner_migrator_v4(*actorv4::MINER_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::POWER_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::POWER_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::SYSTEM_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::SYSTEM_ACTOR_CODE_ID),
+    );
+    migrations.insert(
+        *actorv3::VERIFREG_ACTOR_CODE_ID,
+        nil_migrator_v4(*actorv4::VERIFREG_ACTOR_CODE_ID),
+    );
+
+    // Set of prior version code CIDs for actors to defer during iteration, for explicit migration afterwards.
+    let deferred_code_ids = HashSet::<Cid>::new(); // None in this migration
+
+    if migrations.len() + deferred_code_ids.len() != ACTORS_COUNT {
+        return Err(MigrationError::IncompleteMigrationSpec(migrations.len()));
+    }
+
+    let actors_in = StateTree::new_from_root(store, &actors_root_in).unwrap();
+    let mut actors_out =
+        StateTree::new(store, StateTreeVersion::V3).map_err(MigrationError::StateTreeCreation)?;
+
+    actors_in
+        .for_each(|addr, state| {
+            if deferred_code_ids.contains(&state.code) {
+                return Ok(());
+            }
+
+            let next_input = MigrationJob {
+                address: addr,
+                actor_state: state.clone(),
+                actor_migration: migrations
+                    .get(&state.code)
+                    .cloned()
+                    .ok_or(MigrationError::MigratorNotFound(state.code))?,
+            };
+
+            jobs.push(async move { next_input.run(store, prior_epoch) });
+
+            Ok(())
+        })
+        .map_err(MigrationError::MigrationJobCreate)?;
+
+    task::block_on(async {
```

Comment:
As far as I can tell, none of this needs to be async. `MigrationJob::run` is synchronous and so is `ActorMigration::MigrateState`. I would just accumulate `MigrationJob` in the vector and in the while loop just call `jobs.next().run()`. 

Let me know if I missed some aspect of this.

Reply:
I suppose you want the blocking migrations to be distributed across the threads in the async executor. 

I'm a bit naive on how state migrations get run, could you explain that bit to me?

Reply:
Yeah, i feel the async thing is not of any help here, since all underlying APIs are blocking. Yes, we wanted to distribute each job to run concurrently, (maybe use rayon as an alternative.). But from what i gathered from @ec2 on this, is that the actor state should get updated one after the other but not concurrently. is that correct @ec2 ?

Reply:
As it stands, the actor states will get updated concurrently. In this specific migration, all migrations are nil migrations except one, so distributing it across threads isn't much help anyways. 

Depending on when this is run, the current async setup might not be getting in the way of other tasks either. So, it's probably best to run the functions in a way that communicates the most information to the reader with respect to how these functions fit into the the rest of the logic on the node.

Reply:
So what i had said before was that the StateTree itself can't be used in concurrent settings. You can definitely do everything else concurrently. Its only the set_actor call that probably wont work in a concurrent setting. Concurrency will help even if there is only one migration. In this example, the only migration is for the miner actor. But there are a lot of miner actors.

Reply:
So I've advanced my understanding here with @ec2 :

I was wrong that the concurrency isn't useful here because this routine will run one `MinerMigrator` per miner, and the `block_on` will schedule that task on the Runtime's main thread pool. Because of that, this runs the risk of exhausting the Runtime's main thread pool. A better design would be to schedule this on a separate thread pool for blocking tasks.

Reply:
Yeah, i missed the part that there will be multiple miners that needs to run their migrator. To replace `block_on`, we can go with `task::spawn`

Reply:
I guess we could spawn each job with task::spawn and collect and await on their join handles in `migrate_state_tree`.

Reply:
The main suggestion is to make sure the jobs are scheduled on a thread pool for blocking tasks. Because there are many miners, there will be many blocking tasks which will otherwise exhaust the primary async_std runtime thread pool and block the entire process.